src/bandit-evaluator.ts

```diff
@@ -0,0 +1,178 @@
+    banditModel: BanditModelData,
+  ): BanditEvaluation {
+    const actionScores: Record<string, number> = this.scoreActions(
+      subjectAttributes,
+      actions,
+      banditModel,
+    );
+    const actionWeights: Record<string, number> = this.weighActions(
+      actionScores,
+      banditModel.gamma,
+      banditModel.actionProbabilityFloor,
+    );
+    const selectedActionKey: string = this.selectAction(flagKey, subjectKey, actionWeights);
+    const optimalityGap = 0; // TODO: compute difference between selected and max
+
+    return {
+      flagKey,
+      subjectKey,
+      subjectAttributes,
+      actionKey: selectedActionKey,
+      actionAttributes: actions[selectedActionKey],
+      actionScore: actionScores[selectedActionKey],
+      actionWeight: actionWeights[selectedActionKey],
+      gamma: banditModel.gamma,
+      optimalityGap,
+    };
+  }
+
+  private scoreActions(
+    subjectAttributes: Attributes,
+    actions: Record<string, Attributes>,
+    banditModel: Pick<BanditModelData, 'coefficients' | 'defaultActionScore'>,
+  ): Record<string, number> {
+    const actionScores: Record<string, number> = {};
+    Object.entries(actions).forEach(([actionKey, actionAttributes]) => {
+      let score = banditModel.defaultActionScore;
+      const coefficients = banditModel.coefficients[actionKey];
+      if (coefficients) {
+        score = coefficients.intercept;
+        score += this.scoreNumericAttributes(
+          coefficients.subjectNumericCoefficients,
+          subjectAttributes,
+        );
+        score += this.scoreCategoricalAttributes(
+          coefficients.subjectCategoricalCoefficients,
+          subjectAttributes,
+        );
+        score += this.scoreNumericAttributes(
+          coefficients.actionNumericCoefficients,
+          actionAttributes,
+        );
+        score += this.scoreCategoricalAttributes(
+          coefficients.actionCategoricalCoefficients,
+          actionAttributes,
+        );
+      }
+      actionScores[actionKey] = score;
+    });
+    return actionScores;
+  }
+
+  private scoreNumericAttributes(
+    coefficients: BanditNumericAttributeCoefficients[],
+    attributes: Attributes,
+  ): number {
+    return coefficients.reduce((score, numericCoefficients) => {
+      const attributeValue = attributes[numericCoefficients.attributeKey];
+      if (typeof attributeValue === 'number' && isFinite(attributeValue)) {
+        score += attributeValue * numericCoefficients.coefficient;
+      } else {
+        score += numericCoefficients.missingValueCoefficient;
+      }
+      return score;
+    }, 0);
+  }
+
+  private scoreCategoricalAttributes(
+    coefficients: BanditCategoricalAttributeCoefficients[],
+    attributes: Attributes,
+  ): number {
+    return coefficients.reduce((score, attributeCoefficients) => {
+      const attributeValue = attributes[attributeCoefficients.attributeKey]?.toString();
```

Comment:
If we have a categorical coefficient in our model, and that value is in the context, we'll attempt to convert it to a string if it's not a string. I _think_ this is what we want, to handle things like boolean attributes or (via explicit future passing) numbers treated as categorical values.

Reply:
This is a good point. I think my larger plan is to have the top-level `getBanditAssignment()` method takes attributes that are either separated out or not, similar to our Python SDK ([source](https://github.com/Eppo-exp/python-sdk/blob/main/eppo_client/client.py#L235)). And like that method, only adjust if they are _not_ bucketed, but counting numbers as numeric and everything else as categorical.

At that point, I imagine the logger will get the attribute values as-is. So in the first case it could even get numbers as categorical, and in the second case strings and booleans.

I suppose we'll want to be sure our pipeline `toString()`s it downstream or whatever for categorical, and perhaps defensively ensures its a number before treating it like one for numeric.